src/__spec__/interleave-generator.spec.ts

```diff
@@ -0,0 +1,130 @@
+import assert from 'static-type-assert'
+import * as iter from '../index'
+
+assert<
+  IterableIterator<string>
+>(
+  iter.interleaveGenerator(
+    function * (
+      canTakeAny: Function,
+      buffer: iter.InterleaveBuffer<string>
+    ) {
+      yield buffer.take()
+    }
+  )(['foo'])
+)
```

Comment:
You should add more to this test:

```typescript
assert<
  IterableIterator<string>
>(
  iter.interleaveGenerator(
    function * (
      canTakeAny, // no needs for type here, so that you can test if TypeScript can figure it out (type inference)
      buffer // no type here, same reason as above
    ) {
      // test type of buffer
      assert<iter.InterleaveBuffer<string>>(buffer)

      // expand 'yield buffer.take()'
      const res = buffer.take()
      assert<string>(res) // to test typeof buffer.take()
      yield res // to test if buffer.take() fits in return type

      // add test for buffer.canTake()
      assert<boolean>(buffer.canTake())

      // add test for canTakeAny()
      assert<string>(canTakeAny())
    }
  )(['foo'])
)
```

Reply:
```
Argument of type 'IterableIterator<{}>' is not assignable to parameter of type 'IterableIterator<string>'.
  Type '{}' is not assignable to type 'string'.
```

Reply:
It is inferring `InterleaveBuffer<{}>` for some reason

Reply:
I'll look into this tomorrow (which, to my timezone, is probably until 6-10 hours later).

Reply:
In the meantime, you can explicitly add type annotation to where it causes error.